Fix colored buttons on macOS and Windows

SlySven opened the pull request:
Since somewhere after Qt 5.12 the style code for the "windowssvista" style (only) paints `QPushButton`s with native painting code that hides the background colour set in a stylesheet setting entirely with the border.

This detail is hidden away on: https://doc.qt.io/qt-5/stylesheet-reference.html under QPushButton, where it says:
> "Warning: If you only set a background-color on a QPushButton, the background may not appear unless you set the border property to some value. This is because, by default, the QPushButton draws a native border which completely overlaps the background-color. For example,
> `QPushButton { background-color: red; border: none; }`"

This documentation page does seems to show this from at least back to Qt 5.5 - but it does look that we have not been encountering it - certainly neither the "Fusion" or "Windows" (old style possibly dating back to Win98 era) seem to display this behaviour - which is why this bug fix is labelled as **Windows**.

The fix implemented here uses the `Q_OS_WIN32` to detect that the OS is Windows - and then detects if the style factory being used is the problematic "windowsvista" one (default on Windows nowadays), or something else by appending a `--style=Windows` or `--style=Fusion` to the `Mudlet.exe` command line invokation (possilbly with a short-cut) or does the same thing by setting the `QT_STYLE_OVERRIDE` environmental variable to one of those two styles; if this is the case then a flag is set and that is used to modify some style steet strings to work around the issue.

This ought to be documented in the Lua API functions that might be used to apply stylesheets to pushbuttons as any that set a `background-color` without also doing something with `border` may not behave as the user expects!

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

#### Release post highlight
Fixes a problem with `QPushButton`s not showing a background color when they are supposed to when the (default) `windowsvista` style is in use (WINDOWS OS only).

End-users who are using stylesheets to set just the background color for a QPushButton will also want to take notice of the section for them in https://doc.qt.io/qt-5/stylesheet-reference.html.

SlySven committed:
BugFix: restore QPushButton background colours when windowsvista style used

Since somewhere after Qt 5.12 the style code for the "windowssvista" style
(only) paints `QPushButton`s with native painting code that hides the
background colour set in a stylesheet setting entirely with the border.

This detail is hidden away on:
https://doc.qt.io/qt-5/stylesheet-reference.html
under QPushButton, where it says:
"Warning: If you only set a background-color on a QPushButton, the
background may not appear unless you set the border property to some
value. This is because, by default, the QPushButton draws a native border
which completely overlaps the background-color. For example,

QPushButton { background-color: red; border: none; }"

This documentation page does seems to show this from at least back to Qt
5.5 - but it does look that we have not been encountering it - certainly
neither the "Fusion" or "Windows" (old style possibly dating back to Win98
era) seem to display this behaviour - which is why this bug fix is labelled
as **Windows**.

The fix implemented here uses the `Q_OS_WIN32` to detect that the OS is
Windows - and then detects if the style factory being used is the
problematic "windowsvista" one (default on Windows nowadays), or something
else by appending a `--style=Windows` or `--style=Fusion` to the
`Mudlet.exe` command line invokation (possilbly with a short-cut) or does
the same thing by setting the `QT_STYLE_OVERRIDE` environmental variable to
one of those two styles; if this is the case then a flag is set and that is
used to modify some style steet strings to work around the issue.

This ought to be documented in the Lua API functions that might be used to
apply stylesheets to pushbuttons as any that set a `background-color`
without also doing something with `border` may not behave as the user
expects!

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven commented:
This will probably help with the problem that #4756 was trying to solve and thus close #4743 .

Kebap commented on src/mudlet.cpp:
```diff
@@ -259,6 +260,16 @@ mudlet::mudlet()
+    if (QStyleFactory::keys().contains(QLatin1String("windowsvista"))) {
+        if (qApp->style()->objectName() == QLatin1String("windowsvista")) {
+            mForceWindowsVistaPButtonFix = true;
+            qDebug().nospace().noquote() << "mudlet::mudlet() INFO - \"windowsvista\" has been detected as the style factory in use!";
+        } else {
+            qDebug().nospace().noquote() << "mudlet::mudlet() INFO - \"windowsvista\" has NOT been detected as the style factory in use!";
```
How is this debug output helpful? 😕

SlySven replied:
It also picks up whether the style has been selected *automagically* without user intervention...

Kebap reviewed (approved):
This seems to restore colors to buttons alright!

> This ought to be documented in the Lua API functions that might be used to apply stylesheets to pushbuttons as any that set a `background-color` without also doing something with `border` may not behave as the user expects!

Are you going to then? 😕 

> Release post highlight

None, as not present in 4.10.1

vadi2 reviewed (commented):
Let's see if we can simplify the code - it introduces a lot of complexity for a relatively simple thing.

The same issue seems to affect macOS:

![image](https://user-images.githubusercontent.com/110988/109280208-03e93f80-781b-11eb-83f4-9cd52265ef06.png)

SlySven commented:
I can't speak for the MacOs case as I cannot investigate it. In the past a version of Qt for Mac was missing code to redraw buttons after a state change - this could be more Qt borkage on that platform... is https://bugreports.qt.io/browse/QTBUG-86513 relevant as you are using a dark mode?

SlySven committed:
Revise: simplify previous commit

By making the style sheet creation template strings in the `mudlet`
singleton class we can generate them centrally when the application
starts and provide them globally to all the places where they are needed;
this makes the overall set of changes much smaller!

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven commented on src/mudlet.cpp:
```diff
@@ -259,6 +260,25 @@ mudlet::mudlet()
     loadTranslators(mInterfaceLanguage);
 
+    // In the following the replacement parameter(s) are supplied at points of use:
+#if defined(Q_OS_WIN32)
```
@vadi2 if we want to try something similar for the MacOs case we need to know what the name Qt uses for the MacOs style - if someone can try starting mudlet from a Mac command line with a `-style=Nonsense` argument then on Linux/FreeBSD and indeed Windows the Qt libraries will moan about an invalid style but to get a report about the available ones it can be done by inserting just above here:
```cpp
qDebug().nospace().noquote() << "mudlet::mudlet() INFO - styles available: \n" << QStyleFactory::keys().join(QChar::LineFeed);
```
On my main PC in linux using the system Qt libs (and the Kvantum package installed) I get:
```sh
stephen@Ripley:~/src/mudlet-dev/Mudlet/mudlet/build-mudlet-Desktop_Qt_5_x_GCC8_System-Debug$ ./mudlet -style=Nonsense
QApplication: invalid style override passed, ignoring it.

...
mudlet::mudlet() INFO - styles available: 
Breeze
bb10dark
bb10bright
kvantum-dark
kvantum
cleanlooks
gtk2
cde
motif
plastique
Oxygen
Windows
Fusion

...
```
So if we can do this on a MacOs platform and check what style(s) cause the problem we can try the same sort of fix for JUST that style...

SlySven replied:
Ah, since I left something behind and had to do another commit to fix it I've added that line to the code as well - so we need to see what styles show the problem on Macs...

SlySven committed:
BugFix: remove now undefined (on Windows) variable

Didn't spot it as the last commit was done on a Linux system and the bug
was behind a Windows specific `#if defined(...)`...`#endif`.

Also added a debug line to report the style factories available - to help
in diagnosing a possibly related issue on Macs...

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven commented:
Right - the later commit do simplify things - if we could just test this PR now and identify which styles have a problem on MacOs...

SlySven committed:
Extend: apply same fix to MacOS

Thanks to Discord user `mars#8093` (a.k.a. `The Lost`) for demonstrating
that MacOs seems to have the same problem in it's platform specific
`macintosh` style as well.

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
BugFix: correct some typos in the prior commit

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
Revise: another attempt at fixing MacOS QPushButton coloration

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

vadi2 committed:
Simplify code

vadi2 commented:
The previous code didn't work:

![image](https://user-images.githubusercontent.com/110988/109467142-093cc900-7a6b-11eb-8fe0-22475d455cc9.png)

I've applied some DRY to simplify it

vadi2 commented:
Now works on macOS as well, great job @SlySven:

![Selection_038](https://user-images.githubusercontent.com/110988/109488877-6b56f780-7a86-11eb-96a6-0a6ccfbd9e9f.png)

vadi2 commented on src/mudlet.h:
```diff
@@ -320,6 +319,11 @@ class mudlet : public QMainWindow, public Ui::main_window
 
+    bool firstLaunch = false;
+    // Needed to work around a (likely only Windows) issue:
+    QString mBG_ONLY_BUTTON_SSHEET;
+    QString mFG_BG_BUTTON_SSHEET;
```
The difference between the two - from just looking at their names - isn't clear. Can we improve the name or give better comments?

Kebap replied:
First is a style sheet with background only, second has foreground information, too. 
That's what I get from looking at the names only. Seems fine to me, no?

vadi2 replied:
What's a stylesheet with a background and foreground actually mean?

SlySven replied:
It has a `color: XXX;` entry as well as a `background-color: XXX;` one - and is needed fro those cases where we include text as well as a coloured background on a `QPushButton` so that we can force the text to be a contrasting (White or Black) colour...

SlySven commented:
> Now works on macOS as well, great job @SlySven:

Which commit is that - Discord user "The Lost" was helping me last night with this and it *didn't* work for him - and he was using the `-style macintosh` (and `-style Fusion`) command line arguments to ensure he tested them both...

SlySven commented on src/mudlet.cpp:
```diff
@@ -260,37 +260,17 @@ mudlet::mudlet()
-            mBG_ONLY_BUTTON_SSHEET = QStringLiteral("QPushButton {background-color: %1;}");
-            mFG_BG_BUTTON_SSHEET = QStringLiteral("QPushButton {color: %1; background-color: %2;}");
-        }
+    if (QString stylefactory = qApp->style()->objectName(); QStringList{"windowsvista", "macintosh"}.contains(stylefactory)) {
```
It is all very well trying to avoid repeating oneself but at the time I was NOT sure that the MacOs case had the same solution as the Windows one - you seem to have assume the same will work for both but last time I checked it didn't...

vadi2 replied:
My fixed commit did work, see https://github.com/Mudlet/Mudlet/pull/4846#issuecomment-787862593

SlySven commented on src/mudlet.cpp:
```diff
@@ -260,37 +260,17 @@ mudlet::mudlet()
+        mBG_ONLY_BUTTON_SSHEET = QStringLiteral("QPushButton {background-color: %1; border: 1px solid #8f8f91;}");
+        mFG_BG_BUTTON_SSHEET = QStringLiteral("QPushButton {color: %1; background-color: %2; border: 1px solid #8f8f91;}");
+    } else {
+        qDebug() << "mudlet::mudlet() INFO - " << stylefactory << " is an OK style factory, not applying workarounds";
```
```diff
-qDebug() << "mudlet::mudlet() INFO - " << stylefactory << " is an OK style factory, not applying workarounds";
+qDebug() << "mudlet::mudlet() INFO - '" << stylefactory << "' is an OK style factory, not applying workarounds";
```

SlySven commented on src/mudlet.cpp:
```diff
@@ -260,37 +260,17 @@ mudlet::mudlet()
     loadTranslators(mInterfaceLanguage);
 
-    qDebug().nospace().noquote() << "mudlet::mudlet() INFO - styles available: \n" << QStyleFactory::keys().join(QChar::LineFeed);
+    qDebug().nospace().noquote() << "mudlet::mudlet() INFO - styles available: \n" << QStyleFactory::keys().join(QStringLiteral(", "));
```
```diff
-     qDebug().nospace().noquote() << "mudlet::mudlet() INFO - styles available: \n" << QStyleFactory::keys().join(QStringLiteral(", "));
+     qDebug().nospace().noquote() << "mudlet::mudlet() INFO - styles available: \n'" << QStyleFactory::keys().join(QStringLiteral("', '")) << "'";
```

SlySven committed:
Revise: tweak debug output & make style factory matching case insensitive

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

vadi2 committed:
Add clarity to variable name

vadi2 committed:
Cover the rest as well

vadi2 committed:
Cleanup debug output now that the issue has been solved

vadi2 reviewed (approved):
Confirmed by everyone to work, thanks a lot @SlySven for fixing it :)

SlySven commented:
So, @vadi2, you've taken the reference to the strings being for use with "BUTTONS" out of the name and now only report if the two style factories that are currently problematic are detected.

TBH I would like to keep at least the (revise to) line:
```cpp
 qDebug().nospace().noquote() << "mudlet::mudlet() INFO - '" << stylefactory << "' style factory is believed to not need any workarounds.";
 ```
just so that we can get the style factory that *is* in use should things like this crop up (as I expect they could) whether it is one of these two, or another.

Kebap commented:
```c++
-qDebug().nospace().noquote() << "mudlet::mudlet() INFO - '" << stylefactory << "' style factory is believed to not need any workarounds.";
+qDebug().nospace().noquote() << "mudlet::mudlet() INFO - '" << stylefactory << "' style factory is in use.";
```

> just so that we can get the style factory that _is_ in use

vadi2 commented:
We don't run into this issue often, so printing it will just be spammy most of the time, won't it?

Kebap commented:
Yes we print a lot of spam debug so that's another issue.

vadi2 commented:
Same issue, let's not add to it when we don't really need to.

SlySven commented:
> Same issue, let's not add to it when we don't really need to.

As it happens knowing this detail (given that the style factory for any Qt application CAN be overridden on a per profile or Desktop wide basis) would have made diagnosis quicker as we would not have had to add the lines in and then get testers to confirm which style factory was causing the problem. So a single line giving the one in use - and whether we are doing something to modify the defaults for it - seems a reasonable compromise at the moment. As is reporting, as I am proposing, that we are NOT doing anything - something that we ought to mention in that debugging output so that we do not have to also check that the version is one that has this code in it.

When Qt 6.2 (?) finally gets into FOSS code we will have to reassess whether we still need to do this to get coloured background buttons - given that this seems to run counter to the styling policies of, now 2 mainline OSes...

SlySven commented on src/mudlet.cpp:
```diff
@@ -259,6 +260,15 @@ mudlet::mudlet()
+        mBG_ONLY_STYLESHEET = QStringLiteral("QPushButton {background-color: %1; border: 1px solid #8f8f91;}");
+        mTEXT_ON_BG_STYLESHEET = QStringLiteral("QPushButton {color: %1; background-color: %2; border: 1px solid #8f8f91;}");
+    } else {
+        mBG_ONLY_STYLESHEET = QStringLiteral("QPushButton {background-color: %1;}");
```
```suggestion
        qDebug().nospace().noquote() << "mudlet::mudlet() INFO - '" << stylefactory << "' has been detected as the style factory in use - no styling fixes applied.";
        mBG_ONLY_STYLESHEET = QStringLiteral("QPushButton {background-color: %1;}");
```

Kebap replied:
That is even longer than the last suggestion .. 🙈

SlySven reviewed (commented):
Like this.

vadi2 commented:
OK, let's not spend time on this - I don't see any utility on it but we can always get rid of it later. Add it in and merge the PR. Same should be done for other things like the dictionary loading messages which are just adding a lot of output when nobody has had an issue with them for a year.

SlySven committed:
Update src/mudlet.cpp

Restore a single line to debug output if we do not end up detecting a problematic style factory - it will help us to debug what is going on if end-users have problems with style (factories) and ask us to diagnose things remotely.
